common/lib/capa/capa/tests/test_answer_pool.py

```diff
@@ -353,11 +354,16 @@ def test_answer_pool_2_multiplechoiceresponses_seed1(self):
         problem = new_loncapa_problem(xml_str)
         the_html = problem.get_html()
 
-        str1 = r"<div>.*\[.*'wrong-3'.*'correct-2'.*'wrong-2'.*'wrong-4'.*\].*</div>"
-        str2 = r"<div>.*\[.*'wrong-2'.*'wrong-1'.*'correct-2'.*\].*</div>"    # rng shared
-        # str2 = r"<div>.*\[.*'correct-2'.*'wrong-2'.*'wrong-3'.*\].*</div>"  # rng independent
-
-        str3 = r"<div>\{.*'1_solution_2'.*\}</div>"
+        if six.PY2:
+            str1 = r"<div>.*\[.*'wrong-3'.*'correct-2'.*'wrong-2'.*'wrong-4'.*\].*</div>"
```

Comment:
O.o The ordering changes because of Python3? Is this because it's iterating over a dict without ordering?

My gut reaction is that anything that changes the generated HTML output is going to be dangerous. Do we know why this happens?

Reply:
Cool, that was my reaction as well.  I suspect it's because of dictionaries' default ordering but @Ayub-Khan can you take a look further and provide more info?

Reply:
In taking a quick look at this, I notice that it's not just a different ordering issue.  The choices in the HTML are off-by-1.  In particular, I get the following difference in output for `test_answer_pool_without_solutionset`:

**Python-3:**
**Choices:** 0,2,3,4
```
{'choices': [('choice_4', 'wrong-4'), ('choice_2', 'correct-1'), ('choice_3', 'wrong-3'), ('choice_0', 'wrong-1')], 'value': '', 'input_type': 'radio', 'status': Status('unsubmitted'), 'response_data': {'descriptions': OrderedDict(), 'label': ''}, 'submitted_message': 'Answer received.', 'id': '1_2_1', 'STATIC_URL': '/dummy-static/', 'msg': '', 'name_array_suffix': '', 'describedby_html': Markup('aria-describedby="status_1_2_1"'), 'show_correctness': 'always'}
```

**Python-2**
**Choices:** 1,3,4,5
```
 {'status': Status('unsubmitted'), 'input_type': 'radio', 'STATIC_URL': '/dummy-static/', 'choices': [('choice_3', u'wrong-3'), ('choice_5', u'correct-2'), ('choice_1', u'wrong-2'), ('choice_4', u'wrong-4')], 'submitted_message': 'Answer received.', 'value': '', 'msg': '', 'show_correctness': 'always', 'name_array_suffix': '', 'id': '1_2_1', 'describedby_html': Markup(u'aria-describedby="status_1_2_1"'), 'response_data': {'descriptions': OrderedDict(), 'label': ''}}
```

I'll need to take a deeper look at this.

Reply:
@feanil @ormsbee 
During the creation of LoncapaProblem type problem we [transform](https://github.com/edx/edx-platform/blob/master/common/lib/capa/capa/capa_problem.py#L224) the problem. During the [transformation](https://github.com/edx/edx-platform/blob/master/common/lib/capa/capa/responsetypes.py#L1098) we shuffle the choices using the provided seed value. 
Against a set seed python always produces specific changes. But for different versions those results will be different. As can be seem in screen shots.
**Python2**
![py2](https://user-images.githubusercontent.com/7721119/66034130-0d854f80-e522-11e9-86a7-ac73f80f290e.png)
**Python3**
![py3](https://user-images.githubusercontent.com/7721119/66034153-23931000-e522-11e9-929c-5de489565432.png)
Both of them will produce specific changes no matter. But for results for py2 and py3 wont be same. thats whats happening here.

Reply:
@nasthagiri are there going to be any downstream problems with just mocking out the RNG for the test.  eg. In production, the order produced will still change.  Is that going to impact how the responses get stored or is this just going to affect how the problems are displayed and we're fine with that re-arranging in production?

Reply:
oh right!  Both options do have a potential backward compatibility issue.  I updated my comment to reflect this. 

My assumption is that the randomization should only happen the first time a learner's choices are selected.  The CAPA code should then remember those selections so the learner subsequently sees the exact same ones.  @Ayub-Khan have you verified this and found the code to confirm?

Reply:
My biggest concern would be if changes in randomization would cause state saved in Python 2.7 to become invalid in 3.x. We don't just use random for shuffling multiple choice questions, but also for custom responses. From the bug description, it sounds like the Python seed and rand() behavior remains the same, but the other methods were tweaked to give more even distributions. It might be worth using something the [random2 library](https://pypi.org/project/random2/) to just bring over the Python2 random lib behavior wholesale.

Reply:
@nasthagiri Im not much familiar but during the debugging i only found that we are shuffling once when we are creating the problem. But im not sure if we are saving the state or the seed value and generate state using the seed value every time we show the problem.

Reply:
> My assumption is that the randomization should only happen the first time a learner's choices are selected. The CAPA code should then remember those selections so the learner subsequently sees the exact same ones. @Ayub-Khan have you verified this and found the code to confirm?

My assumption was incorrect. CAPA [stores the seed](https://github.com/edx/edx-platform/blob/2213f524cf8a5f39958e1f67396de4b903b272a1/common/lib/capa/capa/capa_problem.py#L172) and [regenerates](https://github.com/edx/edx-platform/blob/2213f524cf8a5f39958e1f67396de4b903b272a1/common/lib/capa/capa/responsetypes.py#L1279) the choices and shuffles each time.  Hence, we cannot go with either Option 1 or Option 2 - since each results in different randomizations between Python-2 and Python-3.

So, we need to go with @ormsbee's suggestion of using the [random2](https://pypi.org/project/random2/) library - in all cases where the `seed` is stored instead of the generated values.

Reply:
@nasthagiri @feanil @ormsbee https://github.com/edx/edx-platform/pull/21935 using random2 library.

Reply:
Thanks @Ayub-Khan 

I also updated [Python 3 Conversion and Usage Tips](https://openedx.atlassian.net/wiki/spaces/AT/pages/998998951/Python+3+Conversion+and+Usage+Tips) to document this decision.